### PR TITLE
fix: switch to n-api for node@16 support

### PR DESCRIPTION
**Pull request by jbolda**
## Motivation

The current version 1.03 is missing the binary that supports node 15/16. The existing configuration appears to error both on my personal windows and in [CI](https://github.com/sinix-dev/scaffe.js/pull/14/checks?check_run_id=2477607619#step:4:23). It didn't appear to make sense to try to debug what appears to be reliance on old toolchains.

This PR updates to the new version of neon which supports the n-api. The version of n-api used (version 4) currently provides support for node `> 10.17.0` (up to the current v16 at least, and likely future versions) and requires only a single binary that works for all of the versions.

## Approach

Upgrades and reading docs.rs. The neon website seems out of date.

### Alternate Designs

We could have possibly tried to patch what we had, but my previous work with neon and their docs made me feel this was would an effort in vain when we could just switch to something with longer-term node version support.

### Possible Drawbacks or Risks

We are changing the publishing and binaries which is never easy to test perfectly.

We may need to update the n-api version in the future, but this will be much less of a burden than needing a new binding for every version of node. See the [n-api support matrix](https://nodejs.org/api/n-api.html#n_api_node_api_version_matrix).

### TODOs and Open Questions

None known.

## Logs

The CI failure for future reference:
```shell
node-pre-gyp info it worked if it ends with ok
node-pre-gyp info using node-pre-gyp@0.16.0
node-pre-gyp info using node@15.14.0 | win32 | x64
node-pre-gyp WARN Using request for node-pre-gyp https download 
node-pre-gyp info check checked for "D:\a\scaffe.js\scaffe.js\node_modules\ctrlc-windows\dist\index.node" (not found)
node-pre-gyp http GET https://github.com/thefrontside/ctrlc-windows/releases/download/v1.0.3/node-v88-win32-x64.tar.gz
node-pre-gyp http 404 https://github.com/thefrontside/ctrlc-windows/releases/download/v1.0.3/node-v88-win32-x64.tar.gz
node-pre-gyp ERR! install error 
node-pre-gyp ERR! stack Error: 404 status code downloading tarball https://github.com/thefrontside/ctrlc-windows/releases/download/v1.0.3/node-v88-win32-x64.tar.gz
```

The fallback cargo build failure
```shell
error: failed to run custom build command for `neon-sys v0.5.0`

Caused by:
  process didn't exit successfully: `C:\Users\Jacob\Documents\dev\frontside\ctrlc-windows\native\target\debug\build\neon-sys-3cbf1a8623249df2\build-script-build` (exit code: 1)
  --- stdout
  'Skipping node-gyp installation as part of npm install.'
  cargo:node_root_dir=C:\\Users\\Jacob\\.node-gyp\\16.0.0
  cargo:node_lib_file=C:\\Users\\Jacob\\.node-gyp\\16.0.0\\<(target_arch)\\node.lib
  cargo:node_arch=x64
  TARGET = Some("x86_64-pc-windows-msvc")
  HOST = Some("x86_64-pc-windows-msvc")
  AR_x86_64-pc-windows-msvc = None
  AR_x86_64_pc_windows_msvc = None
  HOST_AR = None
  AR = None
  running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.20.27508\\bin\\HostX64\\x64\\lib.exe" "-out:C:\\Users\\Jacob\\Documents\\dev\\frontside\\ctrlc-windows\\native\\target\\debug\\build\\neon-sys-b9fd340399662c31\\out\\libneon.a" "-nologo" "C:\\Users\\Jacob\\Documents\\dev\\frontside\\ctrlc-windows\\native\\target\\debug\\build\\neon-sys-b9fd340399662c31\\out\\native\\build\\Debug\\obj\\neon\\neon.obj"
  LINK : fatal error LNK1181: cannot open input file 'C:\Users\Jacob\Documents\dev\frontside\ctrlc-windows\native\target\debug\build\neon-sys-b9fd340399662c31\out\native\build\Debug\obj\neon\neon.obj'
  exit code: 1181

  --- stderr


  error occurred: Command "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.20.27508\\bin\\HostX64\\x64\\lib.exe" "-out:C:\\Users\\Jacob\\Documents\\dev\\frontside\\ctrlc-windows\\native\\target\\debug\\build\\neon-sys-b9fd340399662c31\\out\\libneon.a" "-nologo" "C:\\Users\\Jacob\\Documents\\dev\\frontside\\ctrlc-windows\\native\\target\\debug\\build\\neon-sys-b9fd340399662c31\\out\\native\\build\\Debug\\obj\\neon\\neon.obj" with args "lib.exe" did not execute successfully (status code exit code: 1181).
```

### CODE REVIEW

File: .changeset/switch-to-n-api.md
Comment (jbolda) [inline .changeset/switch-to-n-api.md]: I think we can maybe have a discussion about this. It is technically just fixing a bug, but this is a big enough change that I could be convinced that we should do a `minor` or a `major` bump.

File: .github/workflows/test.yml
Comment (jbolda) [inline .github/workflows/test.yml]: We are now testing the created binary in all of the major node versions.

File: .github/workflows/upload-binaries.yml
Comment (jbolda) [inline .github/workflows/upload-binaries.yml]: We only need to build one binary and `14` is the highest LTS.

File: package.json
Comment (jbolda) [inline package.json:21]: The way `node-pre-gyp` works with `napi`, it requires a module_path with that `{napi_build_version}`.

File: package.json
Comment (jbolda) [inline package.json]: This package was deprecated for the `@mapbox` one.

File: native/Cargo.toml
Comment (jbolda) [inline native/Cargo.toml:15]: With the updated version, you either use this _or_ the cli. We are using the cli vs `build.js`, so removing this.

File: build.js
Comment (jbolda) [inline build.js:39]: See note on `package.json` (for all of the folder changes).

File: lib/index.js
Comment (jbolda) [inline lib/index.js:4]: See note on `package.json` (for all of the folder changes).

File: package.json
Comment (jbolda) [inline package.json:49]: Match the binary release node version for consistency.

File: native/src/lib.rs
Comment (jbolda) [inline native/src/lib.rs:8]: Line 8 and 9 are the only real changes required. The rest come from rustfmt.

**Review by cowboyd — APPROVED** [review]
@jbolda This is one heckuva PR! Honestly, it looks really great. 

As a safety measure, let's link and run the effection tests on windows with this version. If those are gold, then so are we 👍🏻

File: .github/workflows/upload-binaries.yml
Comment (minkimcello) [reply .github/workflows/upload-binaries.yml]: So with n-api will 14 support 16 too?

File: .github/workflows/test.yml
Comment (minkimcello) [reply .github/workflows/test.yml]: Awesome

**Comment by jbolda** [comment]
Looks like we have green builds!

@minkimcello I presume we merge this one as opposed to #25?

(I will hold off merging just in case.)

File: .github/workflows/upload-binaries.yml
Comment (cowboyd) [reply .github/workflows/upload-binaries.yml]: Yes, that's the nice thing about n-api is that the ABI (Application Binary Interface) is stable , so as new versions are released, we can still rely on the binary interface remaining the same. It's the same concept as semver except at the binary level.

**Comment by cowboyd** [comment]
@jbolda @minkimcello I say we first merge #25  so that 1.0.4 comes out with binaries for node 15 and 16. We know that the legacy binaries work and that our generation process is repeatable. That way, we're guaranteed to have a working solution for the moment while we test out n-api.

I almost think that we should release n-api as a major version upgrade. I know that the JavaScript API is staying the same, but that will still allow us to make this an opt-in upgrade for basically everybody. Thoughts?

**Comment by minkimcello** [comment]
@cowboyd I agree #25 should be merged first. Should I add the matrix of node versions to the test workflow in #25 too? As it's been done in this PR.

**Comment by jbolda** [comment]
> @jbolda @minkimcello I say we first merge #25 so that 1.0.4 comes out with binaries for node 15 and 16. We know that the legacy binaries work and that our generation process is repeatable. That way, we're guaranteed to have a working solution for the moment while we test out n-api.
> 
> I almost think that we should release n-api as a major version upgrade. I know that the JavaScript API is staying the same, but that will still allow us to make this an opt-in upgrade for basically everybody. Thoughts?

I am actually a bit partial to a `major` here as well. Strictly speaking, this is a bug fix => `patch`, but it is a pretty big under the hood change.

With the toolchain errors I saw, I am not confident that rerunning is actually going to produce a binary. The fallback is failing everywhere I have tried / seen. It is worth a shot though especially if we decide to call this a `major`.

**Comment by cowboyd** [comment]
@minkimcello Yeah, let's go ahead and add node 15 and 16 
@jbolda we can always yank the version or try again if the builds fail to generate. We can also try and debug in windows. If we were doing things straight up, we'd actually make sure all the binaries were generated before we tried to publish to NPM at all.

But, lucky for us, there are a great number of integers.

**Review by minkimcello — CHANGES_REQUESTED** [review]
update changeset to change bump and also to include note about how 15/16 is actually not available in 1.0.3?

File: .changeset/switch-to-n-api.md
Comment (jbolda) [inline .changeset/switch-to-n-api.md]:
```suggestion
"ctrlc-windows": major
```

File: .changeset/switch-to-n-api.md
Comment (jbolda) [inline .changeset/switch-to-n-api.md]:
```suggestion
Switch to using the [n-api](https://nodejs.org/api/n-api.html) through the neon bindings. This provides more wholesale support for current and future versions of node, currently v10, v12, v14, and v16. (Previously ctrlc-windows@v1 only provided support for v10, v12 and v14.)
```

**Comment by jbolda** [comment]
@cowboyd since we are doing this as a major, should we create a v2 branch and I can rebase and switch this PR to target that?

**Comment by cowboyd** [comment]
> @cowboyd since we are doing this as a major, should we create a v2 branch and I can rebase and switch this PR to target that?

Absolutely yes, we will need to keep the v1 branch separate since it will be used for any maintenance releases that need to happen there.

**Comment by cowboyd** [comment]
@jbolda We should also make `v2` the default branch

**Comment by jbolda** [comment]
Are we ready to go ahead with those changes? I can take care of that.

**Comment by jbolda** [comment]
@cowboyd added a new v2 branch, targeted this PR towards it, and rebased. I don't have permissions to be able to set v2 as the default branch though.

**Comment by cowboyd** [comment]
@minkimcello did your requested changes get addressed?

**Comment by minkimcello** [comment]
@cowboyd oh yes, it was taken care of.

**Review by minkimcello — APPROVED** [review]

**Comment by cowboyd** [comment]
@jbolda I'll set this to be the default branch once we're merged.

**Comment by jbolda** [comment]
@cowboyd I presume we are ready to merge then?

I think we might need to change this too, but I am not entirely certain _when_ we should: https://github.com/thefrontside/ctrlc-windows/blob/v1/.changeset/config.json#L7

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/thefrontside/ctrlc-windows">thefrontside/ctrlc-windows</a></strong></div><div>Send a CTRL-C event to a Windows console application - thefrontside/ctrlc-windows</div></blockquote>

**Comment by cowboyd** [comment]
This should be part of the merge that creates the first release of 2.0.0 I'm assuming this one?

**Comment by jbolda** [comment]
Alrighty, pushed. Ready to merge then?

**Pull request merged**
